File: backend/learning-service/src/main/java/com/skillspherenexus/learningservice/service/CourseService.java
package com.skillspherenexus.learningservice.service;

import com.skillspherenexus.learningservice.dto.CourseCreateRequestDTO;
import com.skillspherenexus.learningservice.dto.CourseResponseDTO;
import com.skillspherenexus.learningservice.dto.CourseUpdateRequestDTO;
import com.skillspherenexus.learningservice.enums.CourseLevel;
import com.skillspherenexus.learningservice.enums.CourseStatus;
import com.skillspherenexus.learningservice.enums.CourseType;

import java.util.List;
import java.util.UUID;

public interface CourseService {

    CourseResponseDTO createCourse(CourseCreateRequestDTO request);

    CourseResponseDTO updateCourse(
            UUID courseId,
            CourseUpdateRequestDTO request
    );

    CourseResponseDTO getCourseById(UUID courseId);

    CourseResponseDTO getCourseByCode(String courseCode);

    List<CourseResponseDTO> getAllCourses();

    List<CourseResponseDTO> searchCoursesByTitle(String title);

    List<CourseResponseDTO> getCoursesByCategory(String category);

    List<CourseResponseDTO> getCoursesByType(CourseType courseType);

    List<CourseResponseDTO> getCoursesByLevel(CourseLevel courseLevel);

    List<CourseResponseDTO> getCoursesByStatus(CourseStatus status);

    CourseResponseDTO publishCourse(UUID courseId);

    CourseResponseDTO archiveCourse(UUID courseId);

    void deleteCourse(UUID courseId);
}

File: frontend/src/app/learning/services/course.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import { Course } from '../models/course.model';

export interface CourseSaveRequest {
  courseCode?: string;
  title: string;
  description?: string | null;
  category: string;
  courseType: string;
  courseLevel: string;
  pricingType: string;
  price: number;
  currencyCode: string;
  instructorName: string;
  durationHours: number;
  maxCapacity: number;
  passingScore: number;
  certificateEnabled: boolean;
  startDate?: string | null;
  endDate?: string | null;
}

@Injectable({
  providedIn: 'root'
})
export class CourseService {

  private readonly http = inject(HttpClient);

  private readonly apiUrl =
    `${environment.learningApiUrl}/courses`;

  getAllCourses(): Observable<Course[]> {
    return this.http.get<Course[]>(this.apiUrl);
  }

  getCourseById(courseId: string): Observable<Course> {
    return this.http.get<Course>(
      `${this.apiUrl}/${courseId}`
    );
  }

  createCourse(request: CourseSaveRequest): Observable<Course> {
    return this.http.post<Course>(this.apiUrl, request);
  }

  updateCourse(
    courseId: string,
    request: CourseSaveRequest
  ): Observable<Course> {
    return this.http.put<Course>(
      `${this.apiUrl}/${courseId}`,
      request
    );
  }

  deleteCourse(courseId: string): Observable<void> {
    return this.http.delete<void>(
      `${this.apiUrl}/${courseId}`
    );
  }

  publishCourse(courseId: string): Observable<Course> {
    return this.http.patch<Course>(
      `${this.apiUrl}/${courseId}/publish`,
      {}
    );
  }

  archiveCourse(courseId: string): Observable<Course> {
    return this.http.patch<Course>(
      `${this.apiUrl}/${courseId}/archive`,
      {}
    );
  }
}
